src/languages/internal/languagesservice.cpp

```diff
@@ -53,470 +59,340 @@ static QString downloadingStatusTitle()
-        if (!load) {
-            LOGE() << load.toString();
-        }
-        return; // no hash for this language
-    }
-
-    LanguagesHash languageHash = languages().val;
-    if (!languageHash.contains(languageCode)) {
-        LOGE() << "Unknown language: " << languageCode;
-        return;
-    }
+    Language& lang = languageCode == PLACEHOLDER_LANGUAGE_CODE
+                     ? m_placeholderLanguage
+                     : m_languagesHash[effectiveLanguageCode];
 
-    Ret load = loadLanguage(languageCode);
-    if (!load) {
-        LOGE() << load.toString();
-        return;
+    if (!lang.isLoaded()) {
+        loadLanguage(lang);
     }
 
-    Language& language = languageHash[languageCode];
-    m_currentLanguageChanged.send(language);
-}
-
-RetVal<LanguagesHash> LanguagesService::parseLanguagesConfig(const QByteArray& json) const
-{
-    RetVal<LanguagesHash> result;
-
-    QJsonParseError err;
-    QJsonDocument jsonDoc = QJsonDocument::fromJson(json, &err);
-    if (err.error != QJsonParseError::NoError || !jsonDoc.isObject()) {
-        LOGE() << "failed parse, err: " << err.errorString();
-        result.ret = make_ret(Err::ErrorParseConfig);
-        return result;
+    for (QTranslator* t : m_translators) {
+        qApp->removeTranslator(t);
+        delete t;
     }
+    m_translators.clear();
 
-    result.ret = make_ret(Err::NoError);
-    QStringList languages = jsonDoc.object().keys();
-    for (const QString& key : languages) {
-        if (!jsonDoc.object().value(key).isObject()) {
-            continue;
+    for (const io::path_t& file : lang.files) {
+        QTranslator* translator = new QTranslator();
+        bool ok = translator->load(file.toQString());
+        if (ok) {
+            qApp->installTranslator(translator);
+            m_translators << translator;
+        } else {
+            LOGE() << "Error loading translator " << file.toQString();
+            delete translator;
         }
-
-        QJsonObject value = jsonDoc.object().value(key).toObject();
-
-        Language language;
-        language.code = key;
-        language.name = value.value("name").toString();
-        language.archiveFileName = value.value("file_name").toString();
-        language.files = parseLanguageFiles(value);
-
-        result.val.insert(key, language);
     }
 
-    return result;
-}
-
-LanguageFiles LanguagesService::parseLanguageFiles(const QJsonObject& languageObject) const
-{
-    LanguageFiles files;
+    QLocale locale(lang.code);
+    QLocale::setDefault(locale);
+    qApp->setLayoutDirection(locale.textDirection());
 
-    for (const QString& fileType: languageFileTypes) {
-        QJsonObject instrumentsFile = languageObject.value(fileType).toObject();
-        if (!instrumentsFile.isEmpty()) {
-            files << LanguageFile(instrumentsFile.value("file_name").toString(),
-                                  instrumentsFile.value("hash").toString());
+    if (m_inited) {
```

Comment:
As far as I understand, there is no longer any need for this code, since we ask the user to restart the app anyway